syntax = "proto3";

package libctomanage;

option go_package = "github.com/acompany-develop/QuickMPC/proto/libc_to_manage_container";

import "common_types/common_types.proto";
import "google/protobuf/empty.proto";


/**
 * LibcToManage service
*/
service LibcToManage {
    rpc SendShares(SendSharesRequest) returns (google.protobuf.Empty) {}                                    // シェアの送信リクエスト
    rpc DeleteShares(DeleteSharesRequest) returns (google.protobuf.Empty) {}                                // シェアの削除リクエスト
    rpc GetSchema(GetSchemaRequest) returns (GetSchemaResponse) {}                                          // スキーマ取得リクエスト
    rpc ExecuteComputation(ExecuteComputationRequest) returns (ExecuteComputationResponse) {}               // 計算リクエスト
    rpc GetComputationResult(GetComputationRequest) returns (stream GetComputationResultResponse) {}        // 計算結果取得リクエスト
    rpc GetComputationStatus(GetComputationRequest) returns (GetComputationStatusResponse) {}        // 計算Status取得リクエスト
    rpc GetDataList(GetDataListRequest) returns (GetDataListResponse) {}                                    // data一覧取得リクエスト
    rpc GetElapsedTime(GetElapsedTimeRequest) returns (GetElapsedTimeResponse) {}                           // 計算時間取得リクエスト
    rpc GetJobErrorInfo(GetJobErrorInfoRequest) returns (GetJobErrorInfoResponse) {}                        // stacktrace取得リクエスト
    rpc AddShareDataFrame(AddShareDataFrameRequest) returns (AddShareDataFrameResponse) {}                  // テーブル加算リクエスト
}

/**
 * the message of SendSharesRequest
*/
message SendSharesRequest {
    string data_id = 1;
    string shares = 2;
    repeated pb_common_types.Schema schema = 3;
    int32 piece_id = 4;
    string sent_at = 5;
    int32 matching_column = 6;
    string token = 7;
}

/**
 * the message of DeleteSharesRequest
*/
message DeleteSharesRequest {
    repeated string dataIds = 1;
    string token = 2;
}

/**
 * the message of GetSchemaRequest
*/
message GetSchemaRequest {
    string data_id = 1;
    string token = 2;
}

/**
 * the message of GetSchemaResponse
*/
message GetSchemaResponse {
    repeated pb_common_types.Schema schema = 1;
}

/**
 * the message of ExecuteComputationRequest
*/
message JoinOrder{
    repeated string data_ids = 1;
    bool debug_mode = 2;
}
message Input{
    repeated int32 src = 1;
    repeated int32 target = 2;
}
message ExecuteComputationRequest {
    pb_common_types.ComputationMethod method_id = 1;
    string token = 2;
    JoinOrder table = 3;
    Input arg = 4;
}

/*
 * the message of ExecuteComputationResponse
*/
message ExecuteComputationResponse {
    string job_uuid = 3;
}

/**
 * the message of GetComputationResultRequest
*/
message GetComputationRequest {
    string job_uuid = 1;
    string token = 2;
}

/**
 * the message of GetComputationResultResponse
*/
message GetComputationResultResponse {
    repeated string result = 1;
    int32 column_number = 2;
    int32 piece_id = 3;
    oneof result_type{
        bool is_dim1 = 4;
        bool is_dim2 = 5;
        bool is_schema = 6;
    }
}

message GetComputationStatusResponse {
    pb_common_types.JobStatus status = 1;
    optional pb_common_types.JobProgress progress = 2;
}

message GetDataListRequest {
    string token = 1;
}

message GetDataListResponse {
    string result = 1;
}

message GetElapsedTimeRequest {
    string job_uuid = 1;
    string token = 2;
}

message GetElapsedTimeResponse {
    double elapsed_time = 1;
}

message GetJobErrorInfoRequest {
    string job_uuid = 1;
    string token = 2;
}

message GetJobErrorInfoResponse {
    optional pb_common_types.JobErrorInfo job_error_info = 1;
}

message AddShareDataFrameRequest {
    string base_data_id = 1;
    string add_data_id = 2;
    string token = 3;
}

message AddShareDataFrameResponse {
		string data_id = 1;
}
